### PR TITLE
Add a new python script to auto-generate ATen op ODS definitions.

### PR DESCRIPTION
**Pull request by stellaraccident**
* There is still some work on some of the ops to annotate correct types.
* The ODS is not actually included into the dialect yet, but I'd like to commit it so that we can track changes.
* Will reconcile this with the ops produced by the existing script in a followup. I can either attempt to do that in one step or simply include these into the dialect for now and transition all of it over the course of several commits (they do not have colliding names). I'm leaning towards the latter.

### CODE REVIEW

File: include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td [hunk 1]
```diff
@@ -0,0 +1,2117 @@
...
+  let summary = "remainder(input, other, out=None) -> Tensor";
+  let arguments = (ins
+    ATenAnyTensor:$input,
+    ATenAnyTensor:$other
+  );
+  let results = (outs
+    ATenAnyTensor:$result
+  );
+  let description = [{
+    Computes the element-wise remainder of division.
```
Comment (silvasean) [inline include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td:1704]: idea (low priority): If there's a python module that can easily do it, converting these to Markdown would be useful so our ODS doc gen stuff works.

**Review by silvasean — APPROVED** [review]
Looking good. Left some ideas for improvements.

File: include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td [hunk 2]
```diff
@@ -0,0 +1,2117 @@
...
+// Operation summaries and descriptions were systematically derived from public
+// API docstrings and are licensed accordingly:
+//   https://github.com/pytorch/pytorch/blob/master/LICENSE
+//===----------------------------------------------------------------------===//
+// This file is automatically generated.  Please do not edit.
+// Generated via:
+//   python -m npcomp.torch.opdefs.generate_ods 
+//===----------------------------------------------------------------------===//
+
+def ATen_BuiltinAbsOp: ATenImmutableTensorOp<"builtin.abs", [NoSideEffect]> {
```
[the remark this answers is not in the document]
Comment (stephenneuendorffer) [reply include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td:16]: I'd tend to agree.  Is there a reason why these operators need to be distinguished?

File: include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td [hunk 3]
```diff
@@ -0,0 +1,2117 @@
...
+        out (Tensor, optional): the output tensor.
+
+    Example::
+
+        >>> torch.abs(torch.tensor([-1, -2, 3]))
+        tensor([ 1,  2,  3])  
+  }];
+}
+
+def ATen_BuiltinAbsOutrefOp: ATenRefTensorOp<"builtin.abs_outref", []> {
```
[the remark this answers is not in the document]
Comment (stephenneuendorffer) [reply include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td:43]: I'm not a big fan of "Outref".  Maybe "InPlace"?

File: python/npcomp/torch/opdefs/generate_ods.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+      "// See https://llvm.org/LICENSE.txt for license information.",
+      "// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception",
+      "//",
+      "// Operation summaries and descriptions were systematically derived from public",
+      "// API docstrings and are licensed accordingly:",
+      "//   https://github.com/pytorch/pytorch/blob/master/LICENSE",
+      "//===----------------------------------------------------------------------===//",
+      "// This file is automatically generated.  Please do not edit.",
+      "// Generated via:",
+      f"//   python -m {module_name} {' '.join(sys.argv[1:])}",
```
Comment (powderluv) [inline python/npcomp/torch/opdefs/generate_ods.py:167]: nit: this generates a whitespace. 
Screen Shot 2020-09-15 at 11 40 08 PM

File: include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td (hunk 2 above)
[the remark this answers is not in the document]
Comment (stellaraccident) [reply include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td:16]: Good point about the amount of builtin vs non-builtin. If we end up supporting more of the open op registry by way of concrete ODS ops, it makes more sense to qualify those as "custom" or some such thing (vs qualifying all of these as "builtin"). Fwiw - I was primarily trying to create room in the namespace for how it was described to me regarding how PyTorch reasons about these ATen ops.

File: include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td [hunk 5]
```diff
@@ -0,0 +1,2117 @@
...
+    Args:
+        input (Tensor): the input tensor.
+
+    Keyword args:
+        out (Tensor, optional): the output tensor.
+
+    Example::
+
+        >>> torch.abs(torch.tensor([-1, -2, 3]))
+        tensor([ 1,  2,  3])
```
Comment (powderluv) [inline include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td:39]: nit: tensor examples also have whitespaces 
Screen Shot 2020-09-15 at 11 40 26 PM

File: include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td (hunk 3 above)
[the remark this answers is not in the document]
Comment (stellaraccident) [reply include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td:43]: Thanks - I'll switch to InPlace. I was struggling with the name.

File: include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td [hunk 6]
```diff
@@ -0,0 +1,2117 @@
...
+}
+
+def ATen_BuiltinAbsOutrefOp: ATenRefTensorOp<"builtin.abs_outref", []> {
+  let summary = "See variant of op that does not take an out= ref tensor.";
+  let arguments = (ins
+    ATenAnyTensor:$input,
+    ATenAnyRefTensor:$out
+  );
+  let results = (outs
+    ATenAnyRefTensor:$result
```
[the remark this answers is not in the document]
Comment (stellaraccident) [reply include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td:50]: Good catch, thanks. The Python API does indeed return the out ref but there is no reason for these ops to be structured that way.

File: include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td [hunk 7]
```diff
@@ -0,0 +1,2117 @@
...
+        >>> torch.acos(a)
+        tensor([ 1.2294,  2.2004,  1.3690,  1.7298])  
+  }];
+}
+
+def ATen_BuiltinAcosOutrefOp: ATenRefTensorOp<"builtin.acos_outref", []> {
+  let summary = "See variant of op that does not take an out= ref tensor.";
+  let arguments = (ins
+    ATenAnyTensor:$input,
+    ATenAnyRefTensor:$out
```
[the remark this answers is not in the document]
Comment (stellaraccident) [reply include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td:88]: They don't exist yet but need to be in the base td file along with other common predicates and types. Was waiting for some design feedback to wire it up.

**Review by powderluv — COMMENTED** [review]
minor nits.

File: include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td (hunk 1 above)
Comment (stellaraccident) [reply include/npcomp/Dialect/ATen/GeneratedATenBuiltinOps.td:1704]: added todo

**Pull request merged**